dvc/objects/stage.py

```diff
@@ -131,85 +145,151 @@ def _build_tree(path_info, fs, name, odb, state, upload, **kwargs):
+                    return obj
+                except ObjectFormatError:
+                    pass
+    raise FileNotFoundError
+
+
+def _stage_external_tree_info(odb, tree, name):
+    # NOTE: used only for external outputs. Initial reasoning was to be
+    # able to validate .dir files right in the workspace (e.g. check s3
+    # etag), but could be dropped for manual validation with regular md5,
+    # that would be universal for all clouds.
+    assert odb and name != "md5"
 
-    # If we have dir hash in state db, but dir cache file is lost,
-    # then we need to recollect the dir via .get_dir_hash() call below,
-    # see https://github.com/iterative/dvc/issues/2219 for context
-    if (
-        hash_info
-        and hash_info.isdir
-        and not odb.fs.exists(odb.hash_to_path_info(hash_info.value))
-    ):
-        hash_info = None
+    odb.add(tree.path_info, tree.fs, tree.hash_info)
+    raw = odb.get(tree.hash_info)
+    hash_info = get_file_hash(raw.path_info, raw.fs, name, state=odb.state)
+    tree.fs.remove(tree.path_info)
+    tree.path_info = raw.path_info
+    tree.fs = raw.fs
+    tree.hash_info.name = hash_info.name
+    tree.hash_info.value = hash_info.value
+    if not tree.hash_info.value.endswith(".dir"):
+        tree.hash_info.value += ".dir"
+    return tree
 
-    if hash_info:
-        from . import load
-        from .tree import Tree
 
-        obj = load(odb, hash_info)
-        if isinstance(obj, Tree):
-            obj.hash_info.nfiles = len(obj)
-            for key, entry in obj:
-                entry.fs = fs
-                entry.path_info = path_info.joinpath(*key)
-        else:
-            obj.fs = fs
-            obj.path_info = path_info
-        assert obj.hash_info.name == name
-        obj.hash_info.size = hash_info.size
-        return obj
+def stage(
+    odb: Optional["ObjectDB"],
+    path_info: "DvcPath",
+    fs: "BaseFileSystem",
+    name: str,
+    upload: bool = False,
+    **kwargs,
+) -> "HashFile":
+    assert path_info and path_info.scheme == fs.scheme
+
+    details = fs.info(path_info)
+    staging = get_staging(odb, name=name)
+    if odb:
+        try:
+            return _load_from_state(odb, staging, path_info, fs, name)
+        except FileNotFoundError:
+            pass
 
     if details["type"] == "directory":
-        obj = _get_tree_obj(path_info, fs, name, odb, state, upload, **kwargs)
+        obj = _get_tree_obj(
+            path_info, fs, name, odb=odb, upload=upload, **kwargs
+        )
+        if name == "md5":
+            staging.add(obj.path_info, obj.fs, obj.hash_info)
+            raw = staging.get(obj.hash_info)
+            # cleanup unneeded memfs tmpfile and return obj based on staging
+            # ODB fs/path
+            if obj.fs != staging.fs:
+                obj.fs.remove(obj.path_info)
+            obj.fs = raw.fs
+            obj.path_info = raw.path_info
+        else:
+            obj = _stage_external_tree_info(odb, obj, name)
```

Comment:
Staged objects with non-MD5 checksums are still placed directly within the external ODB instead of inside the `/staging` path (so external cache is polluted here). But this is the only way for us to get the checksum for the resulting dir cache file, since checksums like etags are based on location and not file content